File: src/demos/vehicle/test_rigs/demo_VEH_TireTestRig_CRM.cpp
// =============================================================================
// PROJECT CHRONO - http://projectchrono.org
//
// Copyright (c) 2015 projectchrono.org
// All right reserved.
//
// Use of this source code is governed by a BSD-style license that can be found
// in the LICENSE file at the top level of the distribution and at
// http://projectchrono.org/license-chrono.txt.
//
// =============================================================================
// Authors: Radu Serban, Huzaifa Unjhawala
// =============================================================================
//
// Demonstration of the single-wheel tire test rig.
//
// =============================================================================

////#include <float.h>
////unsigned int fp_control_state = _controlfp(_EM_INEXACT, _MCW_EM);

#include <algorithm>

#include "chrono/physics/ChSystemNSC.h"
#include "chrono/physics/ChSystemSMC.h"

#include "chrono_vehicle/ChVehicleDataPath.h"
#include "chrono_vehicle/utils/ChUtilsJSON.h"
#include "chrono_vehicle/wheeled_vehicle/test_rig/ChTireTestRig.h"
#include "chrono_vehicle/wheeled_vehicle/tire/ChDeformableTire.h"
#include "chrono_vehicle/terrain/CRMTerrain.h"

#ifdef CHRONO_VSG
    #include "chrono_fsi/sph/visualization/ChSphVisualizationVSG.h"
#endif

#ifdef CHRONO_POSTPROCESS
    #include "chrono_postprocess/ChGnuPlot.h"
    #include "chrono_postprocess/ChBlender.h"
#endif

#include "chrono_thirdparty/filesystem/path.h"

#include "demos/SetChronoSolver.h"

using namespace chrono;
using namespace chrono::vehicle;
using namespace chrono::fsi;
using namespace chrono::fsi::sph;

using std::cout;
using std::cerr;
using std::endl;

// -----------------------------------------------------------------------------

// Tire specification file
std::string tire_json = "Polaris/Polaris_RigidMeshTire.json";
////std::string tire_json = "Polaris/Polaris_ANCF4Tire_Lumped.json";

// Wheel specification file
std::string wheel_json = "Polaris/Polaris_Wheel.json";

double render_fps = 100;
bool debug_output = false;
bool gnuplot_output = true;
bool blender_output = false;

bool render = true;

// -----------------------------------------------------------------------------

int main() {
    // --------------------------------
    // Create wheel and tire subsystems
    // --------------------------------

    auto wheel = ReadWheelJSON(GetVehicleDataFile(wheel_json));
    auto tire = ReadTireJSON(GetVehicleDataFile(tire_json));

    bool fea_tire = std::dynamic_pointer_cast<ChDeformableTire>(tire) != nullptr;

    // Set tire contact surface (relevant for FEA tires only)
    if (fea_tire) {
        int collision_family = 7;
        auto surface_type = ChTire::ContactSurfaceType::TRIANGLE_MESH;
        double surface_dim = 0;
        tire->SetContactSurfaceType(surface_type, surface_dim, collision_family);
    }

    // ---------------------------------------------------------
    // Create system and set default solver and integrator types
    // ---------------------------------------------------------

    ChSystem* sys = nullptr;
    double step_size = 0;
    ChSolver::Type solver_type;
    ChTimestepper::Type integrator_type;

    if (fea_tire) {
        sys = new ChSystemSMC;
        step_size = 1e-5;
        solver_type = ChSolver::Type::PARDISO_MKL;
        integrator_type = ChTimestepper::Type::EULER_IMPLICIT_LINEARIZED;
    } else {
        sys = new ChSystemNSC;
        step_size = 2e-4;
        solver_type = ChSolver::Type::BARZILAIBORWEIN;
        integrator_type = ChTimestepper::Type::EULER_IMPLICIT_LINEARIZED;
    }

    // Set collision system
    sys->SetCollisionSystemType(ChCollisionSystem::Type::BULLET);

    // Number of OpenMP threads used in Chrono (SCM ray-casting and FEA)
    int num_threads_chrono = std::min(8, ChOMP::GetNumProcs());

    // Number of threads used in collision detection
    int num_threads_collision = 1;

    // Number of threads used by Eigen
    int num_threads_eigen = 1;

    // Number of threads used by PardisoMKL
    int num_threads_pardiso = std::min(8, ChOMP::GetNumProcs());

    sys->SetNumThreads(num_threads_chrono, num_threads_collision, num_threads_eigen);
    SetChronoSolver(*sys, solver_type, integrator_type, num_threads_pardiso);
    tire->SetStepsize(step_size);

    // -----------------------------
    // Create and configure test rig
    // -----------------------------

    ChTireTestRig rig(wheel, tire, sys);

    rig.SetGravitationalAcceleration(9.8);
    rig.SetNormalLoad(2500);

    ////rig.SetCamberAngle(+15 * CH_DEG_TO_RAD);

    rig.SetTireStepsize(step_size);
    rig.SetTireVisualizationType(VisualizationType::COLLISION);

    ChTireTestRig::TerrainParamsCRM params;
    params.radius = 0.01;
    params.mat_props.density = 1700;
    params.mat_props.Young_modulus = 2e6;
    params.mat_props.cohesion_coeff = 1e2;
    params.length = 10;
    params.width = 1;
    params.depth = 0.2;
    rig.SetTerrainCRM(params);

    // -----------------
    // Set test scenario
    // -----------------

    // Scenario: driven wheel
    ////rig.SetAngSpeedFunction(chrono_types::make_shared<ChFunctionConst>(10.0));
    ////rig.Initialize();

    // Scenario: pulled wheel
    ////rig.SetLongSpeedFunction(chrono_types::make_shared<ChFunctionConst>(1.0));
    ////rig.Initialize();

    // Scenario: imobilized wheel
    ////rig.SetLongSpeedFunction(chrono_types::make_shared<ChFunctionConst>(0.0));
    ////rig.SetAngSpeedFunction(chrono_types::make_shared<ChFunctionConst>(0.0));
    ////rig.Initialize();

    // Scenario: prescribe all motion functions
    //   longitudinal speed: 0.2 m/s
    //   angular speed: 10 RPM
    //   slip angle: sinusoidal +- 5 deg with 5 s period
    rig.SetLongSpeedFunction(chrono_types::make_shared<ChFunctionConst>(0.2));
    rig.SetAngSpeedFunction(chrono_types::make_shared<ChFunctionConst>(10 * CH_RPM_TO_RAD_S));
    ////rig.SetSlipAngleFunction(chrono_types::make_shared<ChFunctionSine>(5 * CH_DEG_TO_RAD, 0.2));

    // Scenario: specified longitudinal slip (overrrides other definitons of motion functions)
    ////rig.SetConstantLongitudinalSlip(0.2, 0.1);

    // Set delay before applying inputs (settling time)
    double input_time_delay = 1.0;
    rig.SetTimeDelay(input_time_delay);

    // Initialize the tire test rig
    ////rig.Initialize(ChTireTestRig::Mode::SUSPEND);
    ////rig.Initialize(ChTireTestRig::Mode::DROP);
    rig.Initialize(ChTireTestRig::Mode::TEST);

    // Optionally, modify tire visualization (can be done only after initialization)
    if (auto tire_def = std::dynamic_pointer_cast<ChDeformableTire>(tire)) {
        auto visFEA = chrono_types::make_shared<ChVisualShapeFEA>();
        visFEA->SetFEMdataType(ChVisualShapeFEA::DataType::NODE_SPEED_NORM);
        visFEA->SetShellResolution(3);
        visFEA->SetWireframe(false);
        visFEA->SetColormapRange(0.0, 5.0);
        visFEA->SetSmoothFaces(true);
        tire_def->AddVisualShapeFEA(visFEA);
    }

    // -----------------
    // Initialize output
    // -----------------

    const std::string out_dir = GetChronoOutputPath() + "TIRE_TEST_RIG";
    if (!filesystem::create_directory(filesystem::path(out_dir))) {
        cerr << "Error creating directory " << out_dir << endl;
        return 1;
    }

    // ---------------------------------
    // Create the run-time visualization
    // ---------------------------------

    std::shared_ptr<ChVisualSystem> vis;

#ifdef CHRONO_VSG
    if (render) {
        // FSI plugin
        auto sysFSI = std::dynamic_pointer_cast<CRMTerrain>(rig.GetTerrain())->GetFsiSystemSPH();
        auto visFSI = chrono_types::make_shared<ChSphVisualizationVSG>(sysFSI.get());
        visFSI->EnableFluidMarkers(true);
        visFSI->EnableBoundaryMarkers(true);
        visFSI->EnableRigidBodyMarkers(true);

        // VSG visual system (attach visFSI as plugin)
        auto visVSG = chrono_types::make_shared<vsg3d::ChVisualSystemVSG>();
        visVSG->AttachPlugin(visFSI);
        visVSG->AttachSystem(sys);
        visVSG->SetWindowTitle("Tire Test Rig on CRM deformable terrain");
        visVSG->SetWindowSize(1280, 800);
        visVSG->SetWindowPosition(100, 100);
        visVSG->AddCamera(ChVector3d(1.0, 2.5, 1.0), ChVector3d(0, 1, 0));
        visVSG->SetLightIntensity(0.9f);
        visVSG->SetLightDirection(CH_PI_2, CH_PI / 6);

        visVSG->Initialize();
        vis = visVSG;
    }
#else
    render = false;
#endif

#ifdef CHRONO_POSTPROCESS
    // ---------------------------
    // Create the Blender exporter
    // ---------------------------

    postprocess::ChBlender blender_exporter(sys);

    if (blender_output) {
        std::string blender_dir = out_dir + "/blender";
        if (!filesystem::create_directory(filesystem::path(blender_dir))) {
            cerr << "Error creating directory " << blender_dir << endl;
            return 1;
        }

        blender_exporter.SetBlenderUp_is_ChronoZ();
        blender_exporter.SetBasePath(blender_dir);
        blender_exporter.AddAll();
        blender_exporter.SetCamera(ChVector3d(3, 3, 1), ChVector3d(0, 0, 0), 50);
        blender_exporter.ExportScript();
    }
#endif

    // ---------------
    // Simulation loop
    // ---------------

    // Timers and counters
    ChTimer timer;         // timer for measuring total run time
    double time = 0;       // simulated time
    double sim_time = 0;   // simulation time
    int render_frame = 0;  // render frame counter
    double sim_time_max = 10;

    // Data collection
    ChFunctionInterp long_slip_fct;
    ChFunctionInterp slip_angle_fct;
    ChFunctionInterp camber_angle_fct;

    timer.start();
    while (time < sim_time_max) {
        time = sys->GetChTime();

        if (time >= render_frame / render_fps) {
            auto& loc = rig.GetPos();
            vis->UpdateCamera(loc + ChVector3d(1.0, 2.5, 0.5), loc + ChVector3d(0, 0.25, -0.25));

            if (!vis->Run())
                break;
            vis->Render();
            render_frame++;

#ifdef CHRONO_POSTPROCESS
            if (blender_output)
                blender_exporter.ExportData();
#endif
        }

        rig.Advance(step_size);
        sim_time += sys->GetTimerStep();

                auto long_slip = rig.GetLongitudinalSlip();
        auto slip_angle = rig.GetSlipAngle() * CH_RAD_TO_DEG;
        auto camber_angle = rig.GetCamberAngle() * CH_RAD_TO_DEG;

        if (gnuplot_output && time > input_time_delay) {
            long_slip_fct.AddPoint(time, long_slip);
            slip_angle_fct.AddPoint(time, slip_angle);
            camber_angle_fct.AddPoint(time, camber_angle);
        }

        if (debug_output) {
            cout << time << endl;
            auto long_slip = rig.GetLongitudinalSlip();
            auto slip_angle = rig.GetSlipAngle();
            auto camber_angle = rig.GetCamberAngle();
            cout << "   " << long_slip << " " << slip_angle << " " << camber_angle << endl;
            auto tforce = rig.ReportTireForce();
            auto frc = tforce.force;
            auto pnt = tforce.point;
            auto trq = tforce.moment;
            cout << "   " << frc.x() << " " << frc.y() << " " << frc.z() << endl;
            cout << "   " << pnt.x() << " " << pnt.y() << " " << pnt.z() << endl;
            cout << "   " << trq.x() << " " << trq.y() << " " << trq.z() << endl;
        } else {
            cout << "\rRTF: " << sys->GetRTF();
        }
    }
    timer.stop();

    double step_time = timer();
    cout << "\rSimulated time: " << time << endl;
    cout << "Run time (simulation): " << sim_time << "  |  RTF: " << sim_time / time << endl;
    cout << "Run time (total):      " << step_time << "  |  RTF: " << step_time / time << endl;

#ifdef CHRONO_POSTPROCESS
    // ------------
    // Plot results
    // ------------

    if (gnuplot_output && sys->GetChTime() > input_time_delay) {
        postprocess::ChGnuPlot gplot_long_slip(out_dir + "/tmp1.gpl");
        gplot_long_slip.SetGrid();
        gplot_long_slip.SetLabelX("time (s)");
        gplot_long_slip.SetLabelY("Long. slip");
        gplot_long_slip.SetRangeY(-2, +2);
        gplot_long_slip.Plot(long_slip_fct, "", " with lines lt -1 lc rgb'#00AAEE' ");

        postprocess::ChGnuPlot gplot_slip_angle(out_dir + "/tmp2.gpl");
        gplot_slip_angle.SetGrid();
        gplot_slip_angle.SetLabelX("time (s)");
        gplot_slip_angle.SetLabelY("Slip angle (deg)");
        gplot_slip_angle.SetRangeY(-25, +25);
        gplot_slip_angle.Plot(slip_angle_fct, "", " with lines lt -1 lc rgb'#00AAEE' ");

        postprocess::ChGnuPlot gplot_camber_angle(out_dir + "/tmp3.gpl");
        gplot_camber_angle.SetGrid();
        gplot_camber_angle.SetLabelX("time (s)");
        gplot_camber_angle.SetLabelY("Camber angle (deg)");
        gplot_camber_angle.SetRangeY(-5, +5);
        gplot_camber_angle.Plot(camber_angle_fct, "", " with lines lt -1 lc rgb'#00AAEE' ");
    }
#endif

    return 0;
}

File: src/demos/python/vehicle/demo_VEH_TireTestRig_CRM.py
import os
import math
import pychrono as chrono
import pychrono.vehicle as veh
import pychrono.fsi as fsi
import pychrono.vsg3d as vsg3d
import pychrono.postprocess as postprocess

import importlib.util
from pathlib import Path

# include "demos/SetChronoSolver.h"
file_path = Path(__file__).resolve().parents[1] / "SetChronoSolver.py"

spec = importlib.util.spec_from_file_location("SetChronoSolver", file_path)
SetChronoSolver = importlib.util.module_from_spec(spec)
spec.loader.exec_module(SetChronoSolver)

# -----------------------------------------------------------------------------

# Tire specification file
tire_json = "Polaris/Polaris_RigidMeshTire.json"
# tire_json = "Polaris/Polaris_ANCF4Tire_Lumped.json"

# Wheel specification file
wheel_json = "Polaris/Polaris_Wheel.json"

render_fps = 100
debug_output = False
gnuplot_output = True
blender_output = False

render = True

# -----------------------------------------------------------------------------

def main():
    # --------------------------------
    # Create wheel and tire subsystems
    # --------------------------------
    wheel = veh.ReadWheelJSON(veh.GetVehicleDataFile(wheel_json))
    tire = veh.ReadTireJSON(veh.GetVehicleDataFile(tire_json))

    fea_tire = isinstance(tire, veh.ChDeformableTire)

    # Set tire contact surface (relevant for FEA tires only)
    if fea_tire:
        collision_family = 7
        surface_type = veh.ChTire.ContactSurfaceType_TRIANGLE_MESH
        surface_dim = 0.0
        tire.SetContactSurfaceType(surface_type, surface_dim, collision_family)

    # ---------------------------------------------------------
    # Create system and set default solver and integrator types
    # ---------------------------------------------------------
    sys = None
    step_size = 0
    solver_type = None
    integrator_type = None

    if fea_tire:
        sys = chrono.ChSystemSMC()
        step_size = 1e-5
        solver_type = chrono.ChSolver.Type_PARDISO_MKL
        integrator_type = chrono.ChTimestepper.Type_EULER_IMPLICIT_LINEARIZED
    else:
        sys = chrono.ChSystemNSC()
        step_size = 2e-4
        solver_type = chrono.ChSolver.Type_BARZILAIBORWEIN
        integrator_type = chrono.ChTimestepper.Type_EULER_IMPLICIT_LINEARIZED

    # Set collision system
    sys.SetCollisionSystemType(chrono.ChCollisionSystem.Type_BULLET)

    # Number of OpenMP threads used in Chrono (SCM ray-casting and FEA)
    num_threads_chrono = min(8, chrono.ChOMP.GetNumProcs())

    # Number of threads used in collision detection
    num_threads_collision = 1

    # Number of threads used by Eigen
    num_threads_eigen = 1

    # Number of threads used by PardisoMKL
    num_threads_pardiso = min(8, chrono.ChOMP.GetNumProcs())

    sys.SetNumThreads(num_threads_chrono, num_threads_collision, num_threads_eigen)
    SetChronoSolver.SetChronoSolver(sys, solver_type, integrator_type, num_threads_pardiso)
    tire.SetStepsize(step_size)

    # -----------------------------
    # Create and configure test rig
    # -----------------------------
    rig = veh.ChTireTestRig(wheel, tire, sys)

    rig.SetGravitationalAcceleration(9.8)
    rig.SetNormalLoad(2500)

    rig.SetTireStepsize(step_size)
    rig.SetTireVisualizationType(chrono.VisualizationType_COLLISION)

    size = veh.TerrainPatchSize()
    size.length = 10
    size.width = 1
    size.depth = 0.2

    params = veh.TerrainParamsCRM()
    params.sph_params.initial_spacing = 0.02
    params.mat_props.density = 1700
    params.mat_props.Young_modulus = 2e6
    params.mat_props.cohesion_coeff = 1e2

    rig.SetTerrainCRM(size, params)

    # -----------------
    # Set test scenario
    # -----------------
    rig.SetLongSpeedFunction(chrono.ChFunctionConst(0.2))
    rig.SetAngSpeedFunction(chrono.ChFunctionConst(10 * chrono.CH_RPM_TO_RAD_S))

    input_time_delay = 1.0
    rig.SetTimeDelay(input_time_delay)
    rig.Initialize(veh.ChTireTestRig.Mode_TEST, 0.05)

    # Optionally, modify tire visualization (can be done only after initialization)
    if isinstance(tire, veh.ChDeformableTire):
        tire_def = veh.CastToChDeformableTire(tire)
        visFEA = chrono.ChVisualShapeFEA()
        visFEA.SetFEMdataType(chrono.ChVisualShapeFEA.DataType_NODE_SPEED_NORM)
        visFEA.SetShellResolution(3)
        visFEA.SetWireframe(False)
        visFEA.SetColormapRange(0.0, 5.0)
        visFEA.SetSmoothFaces(True)
        tire_def.AddVisualShapeFEA(visFEA)

    # -----------------
    # Initialize output
    # -----------------
    out_dir = chrono.GetChronoOutputPath() + "TIRE_TEST_RIG"
    if not os.path.exists(out_dir):
        os.makedirs(out_dir)

    # ---------------------------------
    # Create the run-time visualization
    # ---------------------------------
    vis = None
    if render:
        # FSI plugin
        sysFSI = veh.CastToCRMTerrain(rig.GetTerrain()).GetFsiSystemSPH()        
        visFSI = fsi.ChSphVisualizationVSG(sysFSI)

        visFSI.EnableFluidMarkers(True)
        visFSI.EnableBoundaryMarkers(True)
        visFSI.EnableRigidBodyMarkers(True)

        visVSG = vsg3d.ChVisualSystemVSG()
        visVSG.AttachPlugin(visFSI)
        visVSG.AttachSystem(sys)
        visVSG.SetWindowTitle("Tire Test Rig on CRM deformable terrain")
        visVSG.SetWindowSize(1280, 800)
        visVSG.SetWindowPosition(100, 100)
        visVSG.AddCamera(chrono.ChVector3d(1.0, 2.5, 1.0), chrono.ChVector3d(0, 1, 0))
        visVSG.SetLightIntensity(0.9)
        visVSG.SetLightDirection(chrono.CH_PI_2, chrono.CH_PI / 6)

        visVSG.Initialize()
        vis = visVSG

    # ---------------
    # Simulation loop
    # ---------------

    # Timers and counters
    timer = chrono.ChTimer() # timer for measuring total run time
    time = 0.0 # simulated time
    sim_time = 0.0 # simulation time
    render_frame = 0 # render frame counter
    sim_time_max = 10.0

    # Data collection
    long_slip_fct = chrono.ChFunctionInterp()
    slip_angle_fct = chrono.ChFunctionInterp()
    camber_angle_fct = chrono.ChFunctionInterp()

    timer.start()
    while time < sim_time_max:
        time = sys.GetChTime()

        if time >= render_frame / render_fps:
            loc = rig.GetPos()
            vis.UpdateCamera(loc + chrono.ChVector3d(1.0, 2.5, 0.5), loc + chrono.ChVector3d(0, 0.25, -0.25))

            if not vis.Run():
                break
            vis.Render()
            render_frame += 1

        rig.Advance(step_size)
        sim_time += sys.GetTimerStep()

        long_slip = rig.GetLongitudinalSlip()
        slip_angle = rig.GetSlipAngle() * chrono.CH_RAD_TO_DEG
        camber_angle = rig.GetCamberAngle() * chrono.CH_RAD_TO_DEG

        if debug_output and rig.OutputEnabled():
            print(f"{time}")
            print(f"   {long_slip} {slip_angle} {camber_angle}")
            tforce = rig.ReportTireForce()
            frc = tforce.force
            pnt = tforce.point
            trq = tforce.moment
            print(f"   {frc.x} {frc.y} {frc.z}")
            print(f"   {pnt.x} {pnt.y} {pnt.z}")
            print(f"   {trq.x} {trq.y} {trq.z}")
        else:
            print(f"\rRTF: {sys.GetRTF()}", end="")
            
    timer.stop()
    step_time = timer.GetTimeSeconds()
    
    print(f"\nSimulated time: {time}")
    print(f"Run time (simulation): {sim_time}  |  RTF: {sim_time / time if time > 0 else 0}")
    print(f"Run time (total):      {step_time}  |  RTF: {step_time / time if time > 0 else 0}")

if __name__ == "__main__":
    main()